download.go

```diff
@@ -0,0 +1,208 @@
+	u, ok := api.downloadURLCache.Get(id, api.cfg.downloadURLCacheTTL)
+	if ok {
+		return &u, nil
+	}
+
+	return api.updateBundleDownloadURL(ctx, id)
+}
+
+func (api *API) updateBundleDownloadURL(ctx context.Context, id string) (*url.URL, error) {
+	now := time.Now()
+
+	resp, err := apierror.CheckResponse[cluster_api.DownloadBundleResponse](
+		api.cluster.DownloadClusterResourceBundleWithResponse(ctx, id),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("get bundle download URL: %w", err)
+	}
+
+	expiresSeconds, err := strconv.ParseInt(resp.ExpiresInSeconds, 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parse expiration: %w", err)
+	}
+
+	u, err := url.Parse(resp.Url)
+	if err != nil {
+		return nil, fmt.Errorf("parse url: %w", err)
+	}
+
+	api.downloadURLCache.Set(id, *u, now.Add(time.Duration(expiresSeconds)*time.Second))
+	return u, nil
+}
+
+// DownloadResult contains the result of a download operation
+type DownloadResult struct {
+	// NotModified is true if the bundle has not been modified
+	NotModified bool
+	// DownloadConditional contains the new conditional
+	*DownloadConditional
+}
+
+type DownloadConditional struct {
+	ETag         string
+	LastModified string
+}
+
+func (c *DownloadConditional) Validate() error {
+	if c.ETag == "" && c.LastModified == "" {
+		return fmt.Errorf("either ETag or LastModified must be set")
+	}
+	return nil
+}
+
+func (c *DownloadConditional) SetHeaders(req *http.Request) error {
+	if c == nil {
+		return nil
+	}
+	if err := c.Validate(); err != nil {
+		return err
+	}
+	req.Header.Set("If-None-Match", c.ETag)
+	req.Header.Set("If-Modified-Since", c.LastModified)
+	return nil
+}
+
+func newConditionalFromResponse(resp *http.Response) (*DownloadConditional, error) {
+	c := &DownloadConditional{
+		ETag:         resp.Header.Get("ETag"),
+		LastModified: resp.Header.Get("Last-Modified"),
+	}
+	if err := c.Validate(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+type xmlError struct {
+	XMLName xml.Name `xml:"Error"`
+	Code    string   `xml:"Code"`
+	Message string   `xml:"Message"`
+	Details string   `xml:"Details"`
+}
+
+func (e xmlError) Error() string {
+	return fmt.Sprintf("%s: %s", e.Code, e.Message)
+}
+
+func tryXMLError(body []byte) (bool, error) {
+	var xmlErr xmlError
+	err := xml.Unmarshal(body, &xmlErr)
+	if err != nil {
+		return false, fmt.Errorf("unmarshal xml error: %w", err)
+	}
+
+	return true, xmlErr
+}
+
+func httpDownloadError(ctx context.Context, resp *http.Response) error {
+	var buf bytes.Buffer
+	_, err := io.Copy(&buf, io.LimitReader(resp.Body, maxErrorResponseBodySize))
+
+	if isXML(resp.Header.Get("Content-Type")) {
+		ok, err := tryXMLError(buf.Bytes())
+		if ok {
+			return err
+		}
+	}
+
+	log.Ctx(ctx).Debug().Err(err).
+		Str("error", resp.Status).
+		Str("body", buf.String()).Msg("bundle download error")
+
+	return fmt.Errorf("download error: %s", resp.Status)
+}
+
+// isXML parses content-type for application/xml
+func isXML(ct string) bool {
```

Comment:
optional: can we use the `mime` package? https://pkg.go.dev/mime#ParseMediaType

Reply:
right.. I completely forgot it exists